The contact history items are sorted by time (Add support for microseconds to datetime type)

shinde-rahul opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ✔️
| Automated tests included?              | ✔️
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

Events displayed in the **Timeline History** tab on the contact page can occur within the same second, leading to inconsistent ordering. In some cases, events may appear out of sequence simply because all timestamps have identical second-level precision.

This PR introduces microsecond precision to various datetime fields across the application — ensuring tighter ordering and improved reliability for features that depend on accurate event sequencing (such as campaign history).

#### The changes include:
- Updating the `UTCDateTimeType` to support microsecond precision
- Modifying entity mappings to define microsecond precision for applicable datetime columns
- Adding a migration to update existing schema definitions
- Introducing a new functional test to validate correct event ordering when timestamps differ only by microseconds



<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

1. Pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a campaign with two or more actions
3. Trigger the campaign
4. Check an order of events in the Timeline History tab on the contact page

Please see the demo recording, 

https://github.com/user-attachments/assets/1403ec1f-d3d6-4751-92e5-41e4d88dcaec

Note: Pardon my voice.

shinde-rahul committed:
Add fractional seconds to datetime fields used to sort contact history

shinde-rahul committed:
Add migration for the schema change

shinde-rahul committed:
Added coverage

shinde-rahul committed:
Added test coverage

shinde-rahul committed:
Updated the type hints

shinde-rahul committed:
Removed redudant test

shinde-rahul committed:
Checking the with fresh db

shinde-rahul committed:
Testing whether migration is sorted

shinde-rahul committed:
Replaced dump to use the fail

shinde-rahul committed:
Replaced dump to use the fail

shinde-rahul commented:
@escopecz I added this for debugging:

```php
$this->fail(print_r(
    $this->em->getConnection()
        ->executeQuery(sprintf('DESC %slead_points_change_log', MAUTIC_TABLE_PREFIX))
        ->fetchAllAssociative(),
    true
));
```

…inside
`Mautic\CampaignBundle\Tests\Functional\Campaign\ContactCampaignHistoryOrderTest::testContactCampaignHistoryOrderIsCorrectTimeline`.

The test is failing because the migration isn’t kicking in on CI:
[https://github.com/mautic/mautic/actions/runs/19879993362/job/56975589148?pr=15670#step:10:173](https://github.com/mautic/mautic/actions/runs/19879993362/job/56975589148?pr=15670#step:10:173)

Current output:

```
8 => array:6 [
  "Field" => "date_added"
  "Type" => "datetime"
  "Null" => "NO"
  "Key" => "MUL"
  "Default" => null
  "Extra" => ""
]
```

Expected:

```
8 => array:6 [
  "Field" => "date_added"
  "Type" => "datetime(3)"
  "Null" => "NO"
  "Key" => "MUL"
  "Default" => null
  "Extra" => ""
]
```

That says the migration isn’t being executed, which is causing the test failure.

I’m not sure what needs to be done here — need your help.

PS. I also tried running the workflow locally using `act` ([https://github.com/nektos/act](https://github.com/nektos/act)) to reproduce the failure, but couldn’t get `act` working correctly on my machine.

escopecz commented:
@shinde-rahul migrations do not run on new installations. Entity definitions are used for that. So make sure the entities generate the right create table statements.

shinde-rahul committed:
Added SQL declaration to read the presision

shinde-rahul committed:
Added coverage

shinde-rahul committed:
CS fix

shinde-rahul committed:
Removed test debug

shinde-rahul committed:
CS fix

shinde-rahul commented:
PR fixes the issue of failing to update the schema during installation.

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

kniziol commented:
@shinde-rahul I’m not sure how to do this step. Could you please guide me through it?
> 3. Trigger the campaign

I've created the following campaign 👇
| Campaign | Campaign Builder |
| - | - |
| <img width="1486" height="980" alt="Screenshot 2025-12-31 at 17 51 22" src="https://github.com/user-attachments/assets/e284bb73-a9e9-4d0d-8f4d-8724740b4e7c" /> | <img width="1486" height="980" alt="Screenshot 2025-12-31 at 17 51 54" src="https://github.com/user-attachments/assets/e6e8fc59-2fb0-4e1f-9f6c-a535349be855" /> |

kniziol commented on app/bundles/CampaignBundle/Tests/Functional/Campaign/ContactCampaignHistoryOrderTest.php:
```diff
@@ -0,0 +1,164 @@
+        $this->em->persist($contact);
+        $this->em->flush();
+
+        /** @var \Mautic\LeadBundle\Model\ListModel $listModel */
```
```suggestion
        /** @var ListModel $listModel */
```
plus
```php
use Mautic\LeadBundle\Model\ListModel;
```
nitpick: Minor thing

shinde-rahul replied:
Done. Thanks.

kniziol commented on app/bundles/CoreBundle/Doctrine/Mapping/ClassMetadataBuilder.php:
```diff
@@ -243,10 +243,11 @@ public function addPublishDates()
      * @return $this
      */
-    public function addDateAdded($nullable = false)
+    public function addDateAdded($nullable = false, int $precision = 0)
```
```suggestion
    public function addDateAdded(bool $nullable = false, int $precision = 0): static
```
nitpick: Minor thing

kniziol commented on app/migrations/Version20251128024511.php:
```diff
@@ -0,0 +1,76 @@
+        $platform        = $this->connection->getDatabasePlatform();
+        $isMysql         = $platform instanceof MySQLPlatform;
+        $type            = $isMysql ? 'DATETIME' : 'TIMESTAMP(0) WITHOUT TIME ZONE';
+        $nullableColumns = ['date_triggered', 'date_read', 'last_opened', 'date_left'];
```
Lines 61 - 74 are identical to lines 39 - 52:

```php
        $nullableColumns = ['date_triggered', 'date_read', 'last_opened', 'date_left'];

        foreach (self::TABLES_AND_COLUMNS as $table => $columns) {
            $tableName = $this->getPrefixedTableName($table);
            foreach ($columns as $column) {
                if ($isMysql) {
                    $isNullable = in_array($column, $nullableColumns, true);
                    $nullDef    = $isNullable ? 'NULL DEFAULT NULL' : 'NOT NULL';
                    $this->addSql(sprintf('ALTER TABLE %s MODIFY %s %s %s', $tableName, $column, $type, $nullDef));
                } else { // Assuming PostgreSQL
                    $this->addSql(sprintf('ALTER TABLE %s ALTER COLUMN %s TYPE %s', $tableName, $column, $type));
                }
            }
        }
```

suggestion: Could you please avoid this code duplication?

shinde-rahul replied:
I have refactored the migration to dedup. Thanks for the suggestion.

kniziol reviewed (commented):
Please respond to the comments

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
deduplicate date column alteration logic

shinde-rahul committed:
Merge branch 'feature/use-precision-datetime-fields' of github.com:shinde-rahul/mautic into feature/use-precision-datetime-fields

shinde-rahul committed:
Minor nitpick suggestion

Co-authored-by: kniziol <kniziol@users.noreply.github.com>

shinde-rahul committed:
Apply suggestion from @kniziol

Co-authored-by: Krzysztof Nizioł <kniziol@gmail.com>

shinde-rahul committed:
Recor fix

kniziol commented:
# Testing result

✅ Positive

> [!NOTE]
> I’ve successfully tested the changes. The order of campaign actions on the contact history tab is correct. Please refer to the attached screenshots for verification.
> 
> | Campaign actions | Contact history |
> |--------|--------|
> | <img alt="Screenshot 2026-02-08 at 15 04 16" src="https://github.com/user-attachments/assets/43a5d4e9-80b9-4fdd-8dfa-5ac84d0ee114" /> | <img alt="Screenshot 2026-02-08 at 15 06 37" src="https://github.com/user-attachments/assets/f8fdfa6b-b889-4af8-976f-a3c8a6273578" /> | 

> [!TIP]
> If you see the `No Contacts Found` warning on the Contacts tab of the campaign page, as shown in the provided screenshot, you can resolve this issue by running the following command to update the campaign.
>
> <img width="300" alt="Screenshot 2026-02-08 at 13 59 22" src="https://github.com/user-attachments/assets/765966be-ab99-4ca9-bc4f-9bf7977a6783" />
>
> ```bash
> ddev exec console mautic:campaigns:rebuild
> ```
>
> This is important because the campaign cannot be tested without contacts.

@shinde-rahul Great job! Thanks a bunch.

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
fix(report): adjust datetime range parameters for DATETIME(3)

After introducing millisecond precision for datetime columns, report range filters were still binding second-precision values. Update dateFrom/dateTo formatting to Y-m-d H:i:s.u to prevent boundary exclusion in BETWEEN comparisons.

shinde-rahul committed:
Merge branch 'feature/use-precision-datetime-fields' of github.com:shinde-rahul/mautic into feature/use-precision-datetime-fields

shinde-rahul committed:
Fixing the failing tests

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

andersonjeccel reviewed (approved):
100% makes sense, it works

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

escopecz reviewed (commented):
@shinde-rahul I've executed migrations and I can see this migration in the migrations table so it must have executed even though I can't find it in the command output.

But when I check if there are differences between the current DB schema and the schema generated by the entities with `bin/console doctrine:schema:update --dump-sql | grep date` I can see a lot of discrepances:
```
LTER TABLE campaign_lead_event_log CHANGE date_triggered date_triggered DATETIME(3) DEFAULT NULL;
ALTER TABLE email_stats_devices CHANGE date_opened date_opened DATETIME(3) NOT NULL;
ALTER TABLE email_stats CHANGE date_sent date_sent DATETIME(3) NOT NULL, CHANGE date_read date_read DATETIME(3) DEFAULT NULL, CHANGE last_opened last_opened DATETIME(3) DEFAULT NULL;
ALTER TABLE email_stat_replies CHANGE date_replied date_replied DATETIME(3) NOT NULL;
ALTER TABLE form_submissions CHANGE date_submitted date_submitted DATETIME(3) NOT NULL;
ALTER TABLE lead_stages_change_log CHANGE date_added date_added DATETIME(3) NOT NULL;
ALTER TABLE lead_utmtags CHANGE date_added date_added DATETIME(3) NOT NULL;
ALTER TABLE lead_event_log CHANGE date_added date_added DATETIME(3) NOT NULL;
ALTER TABLE lead_donotcontact CHANGE date_added date_added DATETIME(3) NOT NULL;
ALTER TABLE lead_points_change_log CHANGE date_added date_added DATETIME(3) NOT NULL;
ALTER TABLE page_hits CHANGE date_hit date_hit DATETIME(3) NOT NULL, CHANGE date_left date_left DATETIME(3) DEFAULT NULL;
ALTER TABLE video_hits CHANGE date_hit date_hit DATETIME(3) NOT NULL, CHANGE date_left date_left DATETIME(3) DEFAULT NULL;
ALTER TABLE point_lead_action_log CHANGE date_fired date_fired DATETIME(3) NOT NULL;
ALTER TABLE point_lead_event_log CHANGE date_fired date_fired DATETIME(3) NOT NULL;
ALTER TABLE sms_message_stats CHANGE date_sent date_sent DATETIME(3) NOT NULL;
ALTER TABLE stage_lead_action_log CHANGE date_fired date_fired DATETIME(3) NOT NULL;
```

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

andersonjeccel commented on app/bundles/ReportBundle/Tests/Event/ReportGeneratorEventTest.php:
```diff
@@ -332,8 +332,10 @@ public function testApplyFilters(bool $dateOnly, string $condition, string $date
-        yield [false, 't.a_date IS NULL OR (t.a_date BETWEEN :dateFrom AND :dateTo)', 'Y-m-d H:i:s'];
-        yield [true, 't.a_date IS NULL OR (DATE(t.a_date) BETWEEN :dateFrom AND :dateTo)', 'Y-m-d'];
+        return [
+            [false, 't.a_date IS NULL OR (t.a_date BETWEEN :dateFrom AND :dateTo)', 'Y-m-d H:i:s.u'],
```
This method still promises an `\Iterator`, but it returns an array now, so PHPUnit hits a return type error before using the cases. Could we keep the yields or change the return type?

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
refactor(core): apply rector to precision datetime changes

shinde-rahul committed:
test(campaign): fetch lead services by class in history test

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

shinde-rahul committed:
Merge branch '7.x' into feature/use-precision-datetime-fields

hluchas commented on app/bundles/CoreBundle/Doctrine/Type/UTCDateTimeType.php:
Precision should be applied here too to keep consistency or am I wrong?

hluchas commented on app/bundles/CampaignBundle/Tests/Functional/Campaign/ContactCampaignHistoryOrderTest.php:
```diff
@@ -0,0 +1,160 @@
+use Mautic\LeadBundle\Model\ListModel;
+use PHPUnit\Framework\Attributes\DataProvider;
+
+final class ContactCampaignHistoryOrderTest extends MauticMysqlTestCase
```
I'm missig exact coverage for this bugfix here.

- the same Y-m-d H:i:s
- different .u values (e.g. .100, .200, .300)
- proof that the timeline orders by milliseconds, not by event name / ID

hluchas commented on app/bundles/LeadBundle/Entity/TimelineTrait.php:
DateTime 0.1 second later will not be included
```suggestion
                ->setParameter('dateFrom', $options['fromDate']->format('Y-m-d H:i:s.u'))
```

hluchas commented on app/bundles/LeadBundle/Entity/TimelineTrait.php:
```suggestion
                ->setParameter('dateTo', $options['toDate']->format('Y-m-d H:i:s.u'));
```

hluchas commented on app/bundles/LeadBundle/Entity/TimelineTrait.php:
```suggestion
                ->setParameter('dateFrom', $options['fromDate']->format('Y-m-d H:i:s.u'));
```

hluchas commented on app/bundles/LeadBundle/Entity/TimelineTrait.php:
```suggestion
                ->setParameter('dateTo', $options['toDate']->format('Y-m-d H:i:s.u'));
```

hluchas reviewed (changes requested):
Please consider migration of these columns too
 - Asset download - asset_downloads.date_download
 - Integration / sync log - audit_log.date_added
 - Dynamic Content sent - dynamic_content_stats.date_sent

TY!
